File: solutions/java/src/main/java/com/epi/FindMissingAndDuplicateXOR.java
package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class FindMissingAndDuplicateXOR {
  // @include
  private static class DuplicateAndMissing {
    public Integer duplicate;
    public Integer missing;

    public DuplicateAndMissing(Integer duplicate, Integer missing) {
      this.duplicate = duplicate;
      this.missing = missing;
    }
  }

  public static DuplicateAndMissing findDuplicateMissing(List<Integer> A) {
    // Compute the XOR of all numbers from 0 to |A| - 1 and all entries in A.
    int missXORDup = 0;
    for (int i = 0; i < A.size(); ++i) {
      missXORDup ^= i ^ A.get(i);
    }

    // We need to find a bit that's set to 1 in missXORDup. Such a bit
    // must exist if there is a single missing number and a single duplicated
    // number in A.
    //
    // The bit-fiddling assignment below sets all of bits in differBit to 0
    // except for the least significant bit in missXORDup that's 1.
    int differBit = missXORDup & (~(missXORDup - 1));
    int missOrDup = 0;
    for (int i = 0; i < A.size(); ++i) {
      // Focus on entries and numbers in which the differBit-th bit is 1.
      if ((i & differBit) != 0) {
        missOrDup ^= i;
      }
      if ((A.get(i) & differBit) != 0) {
        missOrDup ^= A.get(i);
      }
    }

    // missOrDup is either the missing value or the duplicated entry.
    for (int a : A) {
      if (a == missOrDup) { // missOrDup is the duplicate.
        return new DuplicateAndMissing(missOrDup, missOrDup ^ missXORDup);
      }
    }
    // missOrDup is the missing value.
    return new DuplicateAndMissing(missOrDup ^ missXORDup, missOrDup);
  }
  // @exclude

  private static void simpleTest() {
    List<Integer> A = Arrays.asList(0, 1, 2, 4, 5, 6, 6);
    DuplicateAndMissing dm = findDuplicateMissing(A);
    assert(dm.duplicate == 6 && dm.missing == 3);

    A = Arrays.asList(0, 0, 1);
    dm = findDuplicateMissing(A);
    assert(dm.duplicate == 0 && dm.missing == 2);

    A = Arrays.asList(1, 3, 2, 5, 3, 4);
    dm = findDuplicateMissing(A);
    assert(dm.duplicate == 3 && dm.missing == 0);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = r.nextInt(9999) + 2;
      }
      List<Integer> A = new ArrayList<>(n);
      for (int i = 0; i < n; ++i) {
        A.add(i);
      }
      int missingIdx = r.nextInt(n);
      int missing = A.get(missingIdx);
      int dupIdx = r.nextInt(n);
      while (dupIdx == missingIdx) {
        dupIdx = r.nextInt(n);
      }
      int dup = A.get(dupIdx);
      A.set(missingIdx, dup);
      DuplicateAndMissing ans = findDuplicateMissing(A);
      System.out.println("times = " + times);
      System.out.println(dup + " " + missing);
      System.out.println(ans.duplicate + " " + ans.missing);
      assert(ans.duplicate.equals(dup) && ans.missing.equals(missing));
    }
  }
}

File: static/python/find_missing_and_duplicate_xor.py
import collections
import functools
import sys
import random

# @include
DuplicateAndMissing = collections.namedtuple('DuplicateAndMissing',
                                             ('duplicate', 'missing'))


def find_duplicate_missing(A):
    # Compute the XOR of all numbers from 0 to |A| - 1 and all entries in A.
    miss_XOR_dup = functools.reduce(lambda v, i: v ^ i[0] ^ i[1],
                                    enumerate(A), 0)

    # We need to find a bit that's set to 1 in miss_XOR_dup. Such a bit must
    # exist if there is a single missing number and a single duplicated number
    # in A.
    #
    # The bit-fiddling assignment below sets all of bits in differ_bit
    # to 0 except for the least significant bit in miss_XOR_dup that's 1.
    differ_bit, miss_or_dup = miss_XOR_dup & (~(miss_XOR_dup - 1)), 0
    for i, a in enumerate(A):
        # Focus on entries and numbers in which the differ_bit-th bit is 1.
        if i & differ_bit:
            miss_or_dup ^= i
        if a & differ_bit:
            miss_or_dup ^= a

    # miss_or_dup is either the missing value or the duplicated entry.
    if any(a == miss_or_dup for a in A):
        # miss_or_dup is the duplicate.
        return DuplicateAndMissing(miss_or_dup, miss_or_dup ^ miss_XOR_dup)
    # miss_or_dup is the missing value.
    return DuplicateAndMissing(miss_or_dup ^ miss_XOR_dup, miss_or_dup)
# @exclude


def simple_test():
    A = [0, 1, 2, 4, 5, 6, 6]
    ans = find_duplicate_missing(A)
    assert ans.duplicate == 6 and ans.missing == 3

    A = [0, 0, 1]
    ans = find_duplicate_missing(A)
    assert ans.duplicate == 0 and ans.missing == 2

    A = [1, 3, 2, 5, 3, 4]
    ans = find_duplicate_missing(A)
    assert ans.duplicate == 3 and ans.missing == 0


def main():
    simple_test()
    for times in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(2, 10000)
        A = list(range(n))
        missing_idx = random.randrange(n)
        missing = A[missing_idx]
        dup_idx = random.randrange(n)
        while dup_idx == missing_idx:
            dup_idx = random.randrange(n)
        dup = A[dup_idx]
        A[missing_idx] = dup
        ans = find_duplicate_missing(A)
        print('times =', times)
        print(dup, missing)
        print(ans.duplicate, ans.missing)
        assert ans.duplicate == dup and ans.missing == missing


if __name__ == '__main__':
    main()
